### PR TITLE
Show error message inside downloads popup when download fails because OS issue

### PR DESCRIPTION
**Pull request by jotaemepereira**
Task/Issue URL: https://app.asana.com/0/1204006570077678/1208780258680374/f
Tech Design URL:
CC:

**Description**:
**AC1** 
Given a user on macOS 14.7.1 
When the user fails to download a fail with error code 256 and type NSFileReadUnknownError
Then we show an error banner with some guidance on how to update

**AC2**
Given a user on macOS 15.0.1
When the user fails to download a fail with error code 256 and type NSFileReadUnknownError
Then we show an error banner with some guidance on how to update

**Notes:**
- For the App Store version we should show a `How To Update` button that will take the user to this page: https://support.apple.com/guide/mac-help/get-macos-updates-and-apps-mh35618/mac
For Sparkle users: We will show a `Open Settings` button that it will take them to System Settings → Software Update

non-sandboxed
sandboxed

**Steps to test this PR**:
To test this, you first need to mimic the error. To force this, go to `WebKitDownloadTask`, and after line 330, throw the following exception:

```swift
throw NSError.init(domain: NSCocoaErrorDomain, code: NSFileReadUnknownError)
```

Then, if you are not on macOS 14.7.1 or 15.0.1, you need to add your current OS to the `isSpecificMacOSVersion` function in the `DownloadListViewModel`.

If you want to test Sandbox vs Non-sandbox, you can change how the view is initialized in the `DownloadsViewController` line 140.

**Definition of Done**:

* [x] Does this PR satisfy our [Definition of Done](https://app.asana.com/0/1202500774821704/1207634633537039/f)?

---
###### Internal references:
[Pull Request Review Checklist](https://app.asana.com/0/1202500774821704/1203764234894239/f)
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
[Pull Request Documentation](https://app.asana.com/0/1202500774821704/1204012835277482/f)

### CODE REVIEW

File: DuckDuckGo/FileDownload/Model/DownloadListViewModel.swift [hunk 1]
```diff
@@ -65,6 +67,33 @@ final class DownloadListViewModel {
...
+            }
+            .removeDuplicates()
+            .receive(on: DispatchQueue.main)
+            .sink { [weak self] showError in
+                self?.shouldShowErrorBanner = showError
+            }
+            .store(in: &cancellables)
+    }
+
+    private func isSpecificMacOSVersion() -> Bool {
```
Comment (ayoy) [inline DuckDuckGo/FileDownload/Model/DownloadListViewModel.swift:90]: How about renaming it to e.g. `isAffectedMacOSVersion`, and/or adding a link to Asana task in a comment?

File: DuckDuckGo/FileDownload/View/DownloadsViewController.swift [hunk 2]
```diff
@@ -479,6 +526,52 @@ extension DownloadsViewController: NSTableViewDataSource, NSTableViewDelegate {
...
+    var title: String {
+        switch self {
+        case .openHelpURL: return UserText.downloadsErrorSandboxCallToAction
+        case .openSystemSettings: return UserText.downloadsErrorNonSandboxCallToAction
+        }
+    }
+
+    @MainActor func onAction() {
+        switch self {
+        case .openHelpURL:
```
Comment (ayoy) [inline DuckDuckGo/FileDownload/View/DownloadsViewController.swift:546]: I would suggest that we close downloads popover here (at least in the Sandbox variant where we open a new tab but the popover stays on top).

**Review by ayoy — COMMENTED** [review]
@jotaemepereira thanks for this fix! It works well for me (I had to comment out a few more assert calls) and I only have an autolayout issue when there are more downloads on the list:

Screenshot 2024-11-21 at 08 44 58

Hopefully it's a simple fix.

File: DuckDuckGo/FileDownload/View/DownloadsViewController.swift (hunk 2 above)
Comment (jotaemepereira) [reply DuckDuckGo/FileDownload/View/DownloadsViewController.swift:546]: Good call.

**Comment by jotaemepereira** [comment]
@ayoy Thanks for the feedback! The constraints issue should be fixed:

Screenshot 2024-11-21 at 10 01 31 AM

**Review by ayoy — APPROVED** [review]
LGTM, thanks so much @jotaemepereira!

**Pull request merged**